### ISSUE TITLE
Keras Neural Network Architecture in mlagents

### ISSUE THREAD

**Issue by enikon** [Unity-Technologies/ml-agents#2642]
Dear devs,

I would like to train a model with architecture created using keras.

This is how I tried to achieve that:

First, I wrote keras model :
`i0 = Input(shape=(431,), name='vector_observation')`

`f1 = Dense(8, activation='relu', name='medium')(i0)`
`f2 = Dense(8, activation='relu', name='medium2')(f1)`
`f3 = Dense(8, activation='relu', name='medium3')(f2)`
`fnn = Activation(activation=None, name='action')(f3)`

`model = Model(inputs=i0, outputs=fnn)`
`model.compile(loss='binary_crossentropy', optimizer='Adam', metrics=['binary_accuracy'])`

I froze the model and wrote it to file.pb
then using tensorflow_to_barracuda.py I exported it to file.nn and selected it in the Unity Inspector.

It didn't work, because GenerateChecks() in BarracudaModelParamLoader.cs threw error.

I added to keras :
`f00 = Dense(1, dtype=tf.int32, name='action_output_shape', use_bias=False)`
`model.layers[0].add_weight(name = 'is_continuous_control',shape = [1,],initializer = tf.constant_initializer([1]),trainable = False)`
`model.layers[0].add_weight(name = 'version_number',shape = [1,],initializer = tf.constant_initializer([2]),trainable = False)`
`model.layers[0].add_weight(name = 'memory_size',shape = [1,],initializer = tf.constant_initializer([0]),trainable = False)`
`model.layers[0].add_weight(name = 'action_output_shape',shape = [1,], initializer = tf.constant_initializer([5]),trainable = False)` 

and changed LearningBrain.cs 
`var outp = _engine.Peek(name);` into `var outp = _engine.Peek(name+"/Identity");`
because I needed to have output named 'action' and whatever I did keras decided to rename model.outputs to "action/Identity". ( Maybe you know the way fix that properly.  (?))

Finally, after all that effort Unity ran without error and I was able to inferece the model in file.nn ( I suppose it has worked, model was not trained so outputs were 0 and in AgentAction(float[] vectorAction, ...  ) vectorAction was filled with 0, I might be wrong, however ).

Then I decided to train this model. 

I copied file.nn to /models/test-0 folder and change its name for the name of the trained brain.

Then I tried to start training it with mlagents-learn using --load parameter and appropriate run-id=test.

Unfortunatelly, it failed.

Is there any way to make it work?

My goal is to check if using Dropout layers would decrease the learning time of my agents. I found no way to embed them into model created by ml-agent, so I decided to try that with keras.

Is there any way to achieve that using Unity Environment and mlagents?
Or maybe I'm missing something / doing something wrong ?
Can we force Agent to communicate GetVectorObs and AgentAction with Keras directly ?
I've run out of ideas.

> Help me, Obi-Wan Kenobi. You're my only hope.

**Comment by chriselion** [comment]
Hi @enikon,
Sorry, but we don't support training of arbitrary networks, and we can only support conversion to barracuda for models that were created by ML-Agents.

If you want to try adding dropout to the mode, I'd suggest looking at where we create the graph and making modifications there. For example, here's the code where we create the convolutional layers for visual observations:
https://github.com/Unity-Technologies/ml-agents/blob/99e997c811176b163908f7c619b8aa9947e7e9b7/ml-agents/mlagents/trainers/models.py#L500-L532
